File: 프로그래머스/unrated/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
from heapq import heappop, heappush

def solution(numbers):
    answer = [-1 for i in range(len(numbers))]
    h = []

    for i in range(len(numbers)):
        while h and h[0][0] < numbers[i]:
            _,idx = heappop(h)
            answer[idx] = numbers[i]
        
        heappush(h, [numbers[i],i])
    return answer

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.js
function solution(numbers) {
    const length = numbers.length;
    const answer = Array.from({length: length}, () => -1);
    const stack = [];
    
    for(let i = 0; i < length; i++) {
        while(stack && (numbers[stack.at(-1)] < numbers[i])) {
            answer[stack.pop()] = numbers[i];
        }
        stack.push(i);
    }
    
    return answer;
}